scrapy/http/request/form.py

```diff
@@ -33,7 +33,14 @@ def __init__(self, *args, **kwargs):
                 self.headers.setdefault(b'Content-Type', b'application/x-www-form-urlencoded')
                 self._set_body(querystr)
             else:
-                self._set_url(self.url + ('&' if '?' in self.url else '?') + querystr)
+                url_split = urlsplit(self.url)
+                formdata_key_list = []
+                for k in querystr.split('&'):
+                    formdata_key_list.append(k.split('=')[0])
+                items = []
+                items += [(k, v) for k, v in parse_qsl(url_split.query) if k not in formdata_key_list]
+                query_str = _urlencode(items, self.encoding)
+                self._set_url(urljoin(self.url,'?'+ (query_str + '&' if query_str else '') + querystr + ('#'+ url_split.fragment if url_split.fragment else '')))
```

Comment:
Same here, I think it's better to use stdlib functions for URL construction from parts (see e.g. `urlunsplit` / `urlunparse`)

Reply:
I am not be able to do that, because I have ```query_str``` and ```querystr``` to be appended. We can also have the case where ```query_str can be empty```. So, I explicitly defined in the urljoin. Do you have any suggestion @lopuhin ?

Reply:
> I am not be able to do that, because I have both query_str and querystr to be appended. We can also have the case where query_str can be empty.

@maramsumanth I see, in that case you could still fist concatenate query strings and pass them to urlunsplit, it already has the logic to handle empty parts:

```
In [1]: from urllib.parse import urlunsplit

In [2]: urlunsplit(('a', 'b', 'c', 'd', 'e'))
Out[2]: 'a://b/c?d#e'

In [3]: urlunsplit(('a', 'b', 'c', 'd', ''))
Out[3]: 'a://b/c?d'

In [4]: urlunsplit(('a', 'b', 'c', 'd', ''))
Out[4]: 'a://b/c?d'

In [5]: urlunsplit(('a', 'b', '', '', 'foo'))
Out[5]: 'a://b#foo'
```

Reply:
I thought of using urlunsplit, but both urljoin and urlunsplit yield the same thing.